cmd/collector/collector.go

```diff
@@ -0,0 +1,178 @@
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"net"
+	"os"
+	"os/signal"
+	"strconv"
+	"syscall"
+	"time"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"k8s.io/component-base/logs"
+	"k8s.io/klog"
+
+	"github.com/vmware/go-ipfix/pkg/collector"
+	"github.com/vmware/go-ipfix/pkg/entities"
+	"github.com/vmware/go-ipfix/pkg/registry"
+)
+
+const (
+	logToStdErrFlag = "logtostderr"
+)
+
+var (
+	IPFIXAddr      string
+	IPFIXPort      uint16
+	IPFIXTransport string
+)
+
+func initLoggingToFile(fs *pflag.FlagSet) {
+	var err error
+	var logToStdErr bool
+
+	logToStdErr, err = fs.GetBool(logToStdErrFlag)
+	if err != nil {
+		// Should not happen. Return for safety.
+		return
+	}
+	if logToStdErr {
+		// Logging to files is not enabled.
+		return
+	}
+}
+
+func addIPFIXFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&IPFIXAddr, "ipfix.addr", "", "IPFIX collector address")
+	fs.Uint16Var(&IPFIXPort, "ipfix.port", 4739, "IPFIX collector port")
+	fs.StringVar(&IPFIXTransport, "ipfix.transport", "tcp", "IPFIX collector transport layer")
+}
+
+func printIPFIXMessage(msg *entities.Message) {
+	var buf bytes.Buffer
+	fmt.Fprint(&buf, "\nIPFIX-HDR:\n")
+	fmt.Fprintf(&buf, "  version: %v,  Message Length: %v\n", msg.Version, msg.BufferLength)
+	fmt.Fprintf(&buf, "  Exported Time: %v (%v)\n", msg.ExportTime, time.Unix(int64(msg.ExportTime), 0))
+	fmt.Fprintf(&buf, "  Sequence No.: %v,  Observation Domain ID: %v\n", msg.SeqNumber, msg.ObsDomainID)
+
+	if msg.Set.GetSetType() == entities.Template {
+		fmt.Fprint(&buf, "TEMPLATE SET:\n")
+		for i, record := range msg.Set.GetRecords() {
+			fmt.Fprintf(&buf, "  TEMPLATE RECORD-%d:\n", i)
+			for _, ie := range record.GetInfoElements() {
+				fmt.Fprintf(&buf, "    %s: len=%d (enterprise ID = %d) \n", ie.Element.Name, ie.Element.Len, ie.Element.EnterpriseId)
+			}
+		}
+	} else {
+		fmt.Fprint(&buf, "DATA SET:\n")
+		for i, record := range msg.Set.GetRecords() {
+			fmt.Fprintf(&buf, "  DATA RECORD-%d:\n", i)
+			for _, ie := range record.GetInfoElements() {
+				fmt.Fprintf(&buf, "    %s: %v \n", ie.Element.Name, ie.Value)
+			}
+		}
+	}
+	klog.Infof(buf.String())
+}
+
+func signalHandler(signalCh chan os.Signal, messageReceived chan *entities.Message) {
+	for {
+		select {
+		case msg := <-messageReceived:
+			printIPFIXMessage(msg)
+		case <-signalCh:
+			break
```

Comment:
Wondering whether it is better to use `return` here since `break` will only break the select statement.

Reply:
Yeah. As I was using signalCh directly in main.. it was working fine. `return` should be the proper way. Added a stop channel to use it in main.